yarn-project/archiver/src/archiver/archiver.ts

```diff
@@ -1246,6 +1327,202 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
+   * This is used when the header hash is stale and we can't rely on status() for re-org detection.
+   */
+  private async checkForReorgUsingL2Events(
+    localPendingBlockNumber: number,
+    localPendingBlock: L2Block,
+    fromL1Block: bigint,
+    toL1Block: bigint,
+  ): Promise<boolean> {
+    try {
+      // Get recent L2BlockProposed events
+      const l2BlockProposedLogs = await this.rollup.getContract().getEvents.L2BlockProposed(
+        {},
+        {
+          fromBlock: fromL1Block,
+          toBlock: toL1Block,
+        },
+      );
+
+      if (l2BlockProposedLogs.length === 0) {
+        this.log.debug(
+          `No L2BlockProposed events found for re-org check between L1 blocks ${fromL1Block}-${toL1Block}`,
+        );
+        return true; // Assume no re-org if no events found in recent range
+      }
+
+      // Sort events by block number (ascending) to build chain from genesis
+      const sortedLogs = l2BlockProposedLogs.sort((a, b) => Number(a.args.blockNumber) - Number(b.args.blockNumber));
+
+      // Build a map of block number to most recent event (in case of multiple events for same block)
+      const blockEvents = new Map<number, (typeof sortedLogs)[0]>();
+      for (const log of sortedLogs) {
+        const blockNum = Number(log.args.blockNumber);
+        const existing = blockEvents.get(blockNum);
+        if (!existing || log.blockNumber! > existing.blockNumber!) {
+          blockEvents.set(blockNum, log);
+        }
+      }
+
+      // Find the event for our local pending block
+      const ourBlockEvent = blockEvents.get(localPendingBlockNumber);
+      if (!ourBlockEvent) {
+        this.log.debug(
+          `No L2BlockProposed event found for our block ${localPendingBlockNumber}, checking parent chain`,
+        );
+        return await this.checkParentChainContinuity(localPendingBlockNumber, localPendingBlock, blockEvents);
+      }
+
+      // Check if our local block's header hash matches what's on L1
+      const ourLocalHeaderHash = localPendingBlock.header.toPropose().hash().toString();
+      const l1HeaderHash = ourBlockEvent.args.headerHash;
+
+      if (ourLocalHeaderHash !== l1HeaderHash) {
+        this.log.warn(`L2 re-org detected: header hash mismatch for block ${localPendingBlockNumber}`, {
+          localPendingBlockNumber,
+          ourLocalHeaderHash,
+          l1HeaderHash,
+          l1BlockNumber: ourBlockEvent.blockNumber,
+        });
+        return false;
+      }
+
+      // Verify parent chain continuity
+      return await this.checkParentChainContinuity(localPendingBlockNumber, localPendingBlock, blockEvents);
+    } catch (error) {
+      this.log.warn(`Error checking for re-org using L2 events, assuming re-org occurred`, { error });
+      return false; // Conservative: assume re-org if we can't check
+    }
+  }
+
+  /**
+   * Verifies that the parent chain is consistent by checking parent header hashes.
```

Comment:
Checks that the current chain have not been pruned and is consistent with the incoming blocks by looking at the parent header hashes of the incoming blocks.

Could you not directly have that the `blockEvents` include a second event for the `localPendingblock`. that would be caught though so no biggie.